### ISSUE TITLE
Windows Bcrypt API requires output buffer length

### ISSUE THREAD

**Issue by huitema** [h2o/picotls#273]
As mentioned in issue #271 I am trying to write a crypto provider for Windows implementations. I have an issue with the length of coded or decoded fields. For example, the decrypt API on WIndows is defined as:
~~~
NTSTATUS BCryptDecrypt(
  BCRYPT_KEY_HANDLE hKey,
  PUCHAR            pbInput,
  ULONG             cbInput,
  VOID              *pPaddingInfo,
  PUCHAR            pbIV,
  ULONG             cbIV,
  PUCHAR            pbOutput,
  ULONG             cbOutput,
  ULONG             *pcbResult,
  ULONG             dwFlags
);
~~~
Note the three fields defining the output: a byte pointer to the output buffer `pbOutput`, the length of the output buffer in bytes `cbOutput`, and the length of the decoded result `*pcbResult`. This does not match well with the Picotls API:
~~~
typedef struct st_ptls_aead_context_t {
    const struct st_ptls_aead_algorithm_t *algo;
    uint8_t static_iv[PTLS_MAX_IV_SIZE];
    void (*dispose_crypto)(struct st_ptls_aead_context_t *ctx);
    void (*do_encrypt_init)(struct st_ptls_aead_context_t *ctx, const void *iv, const void *aad, size_t aadlen);
    size_t (*do_encrypt_update)(struct st_ptls_aead_context_t *ctx, void *output, const void *input, size_t inlen);
    size_t (*do_encrypt_final)(struct st_ptls_aead_context_t *ctx, void *output);
    size_t (*do_decrypt)(struct st_ptls_aead_context_t *ctx, void *output, const void *input, size_t inlen, const void *iv,
                         const void *aad, size_t aadlen);
} ptls_aead_context_t;
~~~
In these API, the application just provides `void *output` without specifying a buffer size. How would you fill up parameters like `cbOutput` ?

**Comment by kazuho** [comment]
That API of picotls is specific to AEAD, therefore the size of the output is always the size of the input minus AEAD tag size.

**Comment by huitema** [comment]
That API, yes. But then we also have:
~~~
typedef struct st_ptls_cipher_context_t {
    const struct st_ptls_cipher_algorithm_t *algo;
    void (*do_dispose)(struct st_ptls_cipher_context_t *ctx);
    void (*do_init)(struct st_ptls_cipher_context_t *ctx, const void *iv);
    void (*do_transform)(struct st_ptls_cipher_context_t *ctx, void *output, const void *input, size_t len);
} ptls_cipher_context_t;
~~~
I will need this for `AES128-ECB`, `AES128-CTR`, `AES256-ECB` and`AES256-CTR`. What is the output length there? Do I have to compute it from the block size, or do we assume that the input length is always a multiple of the block size?

**Comment by kazuho** [comment]
That's a good point, and I should clarify that in the doc-comment.

The contract of the API is that the input passed to `ptls_cipher_encrypt` must be a multiple of the block size (or multiple of 1 in case of CTR).

**Comment by huitema** [comment]
OK. I see that CTR is not in fact implemented in Windows. I would have to do it myself, encrypting IV^Counter using ECB and then XOR the result with the input.

**Comment by huitema** [comment]
Fixed in PR #275
